src/mds/Server.cc

```diff
@@ -1242,6 +1251,15 @@ void Server::evict_cap_revoke_non_responders() {
   }
 
   auto&& to_evict = mds->locker->get_late_revoking_clients(cap_revoke_eviction_timeout);
+  // don't kick client if ops or osds are laggy
+  if(to_evict.size() && mds->objecter->with_osdmap([](const OSDMap &map)
+  { return map.are_osds_laggy(); })) {
+    mds->clog->warn() << "laggy OSDs deteced, skipping laggy clients eviction "
+                      << "until they return to normal";
```

Comment:
The clog warning needs removed because the MDS will spam the log every few seconds. Use the health warnings via the MDS beacons instead.

Reply:
do you mean to use `MDSHealthMetric` from `MMDSBeacon`?

Reply:
Yes

Reply:
do you want me to just write them down as part of `Beacon::notify_health` thus making `MDSRankDispatcher::tick` trigger the warning or explicitly push the warning into health metrics vector thro' `Server::find_idle_sessions` and `Server::evict_cap_revoke_non_responders`?

Reply:
> do you want me to just write them down as part of Beacon::notify_health thus making MDSRankDispatcher::tick trigger the warning

Yes. You can add introspection in `Server` which keeps track of the number of clients which are not evicted because an OSD is laggy. You'll need to change how the logic works in this method to count the number of sessions (a "dry-run" of the search for idle sessions).

Reply:
> You'll need to change how the logic works in this method to count the number of sessions (a "dry-run" of the search for idle sessions)

If we want to have a dry runs of such idle sessions eligible for eviction then i think we can just let the vector `to_evict` get filled up in `find_idle_sessions()` and check for lagginess just before eviction i.e. here https://github.com/ceph/ceph/blob/main/src/mds/Server.cc#L1228

here I'm thinking of someting like 
```
if (to_evict.size() && laggy) {
	beacon.notify_lagginess(to_evict.size()); // notify_lagginess a new method in Beacon
}```

same thing can also be implemented in `evict_cap_revoke_non_responders()` which records all the non responders in a vector

Reply:
It looks like an okay approach.

Reply:
is it important to count the number of idle sessions? what if we just send the message that laggy OSDs are making clients unresponsive?